lsp-mode.el

```diff
@@ -8356,8 +8391,234 @@ See https://github.com/emacs-lsp/lsp-mode."
+
+(declare-function org-babel-get-src-block-info "ext:org")
+(declare-function org-do-remove-indentation "ext:org")
+(declare-function org-src-get-lang-mode "ext:org")
+(declare-function org-element-context "ext:org")
+
+(defun lsp--virtual-buffer-update-position ()
+  (-if-let (virtual-buffer (-first (-lambda ((&plist :in-range))
+                                     (funcall in-range))
+                                   lsp--virtual-buffer-connections))
+      (unless (equal virtual-buffer lsp--virtual-buffer)
+        (lsp-org))
+    (when lsp-managed-mode
+      (lsp-managed-mode -1)
+      (lsp-mode -1)
+      (setq-local lsp--buffer-workspaces nil)
+      (setq-local lsp--virtual-buffer nil)
+      (setq-local lsp-buffer-uri nil)
+
+      ;; force refresh of diagnostics
+      (run-hooks 'lsp-after-diagnostics-hook))))
+
+(defun lsp-virtual-buffer-on-change (start end length)
+  "Adjust on change event to be executed against the proper language server."
+  (let ((max-point (max end
+                        (or (plist-get lsp--before-change-vals :end) 0)
+                        (+ start length))))
+    (when-let (virtual-buffer (-first (lambda (vb)
+                                        (let ((lsp--virtual-buffer vb))
+                                          (and (lsp-virtual-buffer-call :in-range start)
+                                               (lsp-virtual-buffer-call :in-range max-point))))
+                                      lsp--virtual-buffer-connections))
+      (lsp-with-current-buffer virtual-buffer
+        (lsp-on-change start end length
+                       (lambda (&rest _)
+                         (list :range (lsp--range (list :character 0 :line 0)
+                                                  lsp--virtual-buffer-point-max)
+                               :text (lsp--buffer-content))))))))
+
+(defvar-local lsp--virtual-buffer-point-max nil)
+
+(defun lsp-virtual-buffer-before-change (start _end)
+  (when-let (virtual-buffer (-first (lambda (vb)
+                                      (lsp-with-current-buffer vb
+                                        (lsp-virtual-buffer-call :in-range start)))
+                                    lsp--virtual-buffer-connections))
+    (lsp-with-current-buffer virtual-buffer
+      (setq lsp--virtual-buffer-point-max
+            (lsp--point-to-position (lsp-virtual-buffer-call :last-point))))))
+
+(defun lsp-patch-on-change-event ()
+  (remove-hook 'after-change-functions #'lsp-on-change t)
+  (add-hook 'after-change-functions #'lsp-virtual-buffer-on-change nil t)
+  (add-hook 'before-change-functions #'lsp-virtual-buffer-before-change nil t))
+
+(defun lsp-kill-virtual-buffers ()
+  (mapc #'lsp-virtual-buffer-disconnect lsp--virtual-buffer-connections))
+
+(defun lsp--move-point-in-indentation (point indentation)
+  (save-excursion
+    (goto-char point)
+    (if (<= point (+ (point-at-bol) indentation))
+        (point-at-bol)
+      point)))
+
+(defun lsp-org ()
```

Comment:
Can we move this into another file? `lsp-org.el` for example.
I imagine that we would have another one for `markdown-mode` too

Reply:
Another thought, since this function can be applied even for something not `org-mode`? Should we rename this to something more generic and passing a parameter / object that provides `(org-element-context)`, `(cl-third (org-babel-get-src-block-info 'light))` ... as generic functions or properties?
Then creating `markdown-mode` integration can be as simple as define those interfaces/functions

Reply:
Once we settle down and this is tested enough my plan is to split `lsp-org` into 2 parts: common and mode-specific. The mode-specific will be building a `virtual-buffer` and pass to the common one. Then we could move lsp-org in a separate file if it is more than a dozen lines.

The issue with this PR is that I have touched a lot of places and I am constantly having to merge and reimplement once the master branch changes. I want to have it in and then focus on converting the code to use lsp-protocol.el. Once we move to plists and lsp-protocol.el and receive enough feedback for org-mode integration we could start addressing your comments.

WDYT?

Reply:
Sure